### ISSUE TITLE
TOUCH command

### ISSUE THREAD

**Issue by xandark** [burniouf/A2osX#120]
Using the touch command is a common shell scripting idiom to create a file if one doesn't exist. Would it be a good idea to implement this in a simple way that it either creates a file, or updates the timestamp of one if it already exists?

**Comment by patrickbst** [comment]
Xandark, you can create an empty file (i.e. myfile) by using the command **ECHO -N "" > myfile**.  This will create a zero byte file with the current date and time (assuming a supported clock is present).  If you want to simulate "touching" a file to update its date, you can copy the file to a tmpfilename and then copy that file back to the original, this preserves the original creation date and updates the mod date based on the time of the copy.  For fun you could combine this all into a nifty TOUCH script, an example of which I have placed in the EXAMPLES folder.  Of course you could do the same logic to make your own TOUCH BIN/External command in Assembler.  If you wanted to do this Xandark, I would suggest starting with the source to CHTYP which already modifies the file attributes of a file, but instead of changing the Type Bits, set the mod date bits to 00000 and Prodos will automatically set the mod date to the current date and time.  This would be a good first project to introduce you to make A2osX external commands of your own.

**Comment by xandark** [comment]
Hello Patrick,

Yes, I was also musing of echoing out a null string to a file as an alternative, glad you made a script for it. I looked at it and it's certainly a work-around, though I would grit my teeth if I had to use it with its copying of files on a slow ProDOS file system.

What I love is that your script already speaks to the power and versatility of A2osX :)

Very interesting and thanks for the tip for where to start with creating the command. I cannot implement it at this time in my life, I have too many obligations but I do like to see how A2osX is doing. I suggest making a new file that has a list of "junior jobs" that people could pick up to get into the project. And definitely add this to the list, it would be perfect.

**Comment by burniouf** [comment]
I see TOUCH command is mandatory to go further with A2osX....
Just for my personal knowledge, may i could ask why it prevent testing/playing with any other stuff already working is SH, and more generally in A2osX ?

It will take few minutes to make TOUCH written, but at least, could you provide specs ? (switches...etc...)

Due to limited access to file attibutes in STANDARD ProDOS 2.0.3, i fear most of interesting switches (access to CREATION time.....) will rely on ProDOS.FX.

**Comment by patrickbst** [comment]
TOUCH command specification
TOUCH [OPTION]... FILESPEC... where FILESPEC is a single file name or partial filename with wildcards, * alone would be all files in the current or specified directory, and  OPTION is one or more of the following:
-c : do not create a file if it does not exist
-d DATE : use the date in the form DD/MM/YY as the date to use when touching files
-m : change only the modification date, not the creation date
-t TIME : use the time in the format YYMMDDhhmm as the time when touching files
-h : display help/usage

Examples:
TOUCH *   # touch all files with current date and time
TOUCH -D 25/12/19 foo* . # touch all files whose names begin with foo in the current directory, setting the date to xmas
TOUCH -m -t 1912251510 /MYVOL/MYDIR/xmas$card # touch the files on the drive MYVOL in the directory MYDIR whose name is 9 chars long and begins with xmas and ends with card and for these files update only the modification date and set it to exactly christmas dat at 3 10 in the afternoon.

**Comment by xandark** [comment]
I agree with your suggestions but this one is counterintuitive for me:

> -M : change only the modification date, not the creation date

The way that I am expecting this to work is that by default TOUCH  changes the modification date, not the creation date. This is how the GNU TOUCH command works. I'd leave this switch out.

**Comment by patrickbst** [comment]
All of the doc i looked at suggests that by default the TOUCH command updates **both** the creation and access date/times.  For prodos there is no access time, so only create would be touched by default.  To update the modify time you use **-m**, which is what i put to follow the standard.  There is another option **-a** which is update access time only that we may in the future what to add if other O/S volumes are supported such as though mounted file systems.
